### ISSUE TITLE
Implement ForkJoinDispatcher

### ISSUE THREAD

**Issue by Aaronontheweb** [akkadotnet/akka.net#675]
The `ForkJoinDispatcher` is pretty straightforward - it's a configurable dispatcher that runs on top of a dedicated number of threads. Pretty similar to what the `DedicatedThreadPoolFiber` does in Helios: https://github.com/Aaronontheweb/helios/blob/master/src/Helios/Concurrency/Impl/DedicatedThreadPoolFiber.cs#L13

There are a number of internal system-level dispatcher that depend on this, such as the dedicated remoting and clustering dispatchers.

**Comment by rogeralsing** [comment]
@VesaKarvonen  that makes the Hopac lib have agreed to help us optimize our dispatchers.
So jump into the Hopac gitter chan and talk to him :)

**Comment by VesaKarvonen** [comment]
Or we can talk on the Akka.net gitter channel. :)

**Comment by Aaronontheweb** [comment]
Hi @VesaKarvonen ! Sorry for not responding sooner - super interested in your work on Hopac and what your experience can do to improve both Akka.NET and Helios. What time of day is best to reach you on the Gitter?

**Comment by Aaronontheweb** [comment]
Created an initial spec for this: https://gist.github.com/Aaronontheweb/97bd3dcf0cd4f425f464

**Comment by Aaronontheweb** [comment]
So I spent the past couple of days working on this: https://github.com/helios-io/DedicatedThreadPool

Uses core CLR code to do work-stealing and a lot of the other fun stuff that comes straight from the CLR threadpool.

Unfortunately, its performance absolutely sucks.

DedicatedThreadPool Benchmark
-----
Asus g70 gaming laptop - 4 cores Intel i7 @2.4ghz

Comparing Helios.Concurrency.DedicatedThreadPool vs System.Threading.ThreadPool for 10000 items
DedicatedThreadPool.NumThreads: 8
ETW Logging NOT Enabled
System.Threading.ThreadPool
00:00:00.0040000
Helios.Concurrency.DedicatedThreadPool
00:00:02.4780000

I had to cut the benchmark off at 10000 items, because the latency became simply hilarious any further along than that.

The thing I realized is that work-stealing queues only offer performance benefits for recursive thread queueing (aka `ContinueWith`), not so much for traditional queueing. Might work well inside Akka.NET when actors on the same `ForkJoinDispatcher` message each other, but won't have an impact at all otherwise.

The other issue with this design is that although it lazily creates / eliminates threads, that overhead really adds up on work queue misses - which happen often. I added some ETW traces and showed that for every queue hit there are ~2 misses. Means there's a lot of threads that do at most 1 piece of work before exiting, which is terrible.

I tried a design where threads waited for work on a `ManualResetEvent` / `AutoResetEvent` instead of exiting and the performance was even worse, with more CPU consumption.

TL; DR; - using the CLR's design as-is ain't gonna cut it, even though work stealing queues are pretty cool.  Without having access to the low-level code in the CLR for reusing Win32 thread primitives you end up wasting a lot of resources.

TL;DR; 2 - there's a fair chance I might have really screwed something up when I was trying to follow the CLR implementation. Please take a look at https://github.com/helios-io/DedicatedThreadPool and send a PR if something's off :p 

DedicatedThreadFiber Benchmark
-----
After having my heart broken by the DedicatedThreadPool, I decided to see how the original Helios.Concurrency.DedicatedThreadFiber compared.

Comparing Helios.Concurrency.DedicatedThreadPool vs Helios.Concurrency.DedicatedThreadFiber for 10000 items
DedicatedThreadFiber.NumThreads: 8
System.Threading.ThreadPool
00:00:00.0040000
Helios.Concurrency.DedicatedThreadFiber
00:00:00.0140000
Comparing Helios.Concurrency.DedicatedThreadPool vs Helios.Concurrency.DedicatedThreadFiber for 100000 items
DedicatedThreadFiber.NumThreads: 8
System.Threading.ThreadPool
00:00:00.0460000
Helios.Concurrency.DedicatedThreadFiber
00:00:00.1990000
Comparing Helios.Concurrency.DedicatedThreadPool vs Helios.Concurrency.DedicatedThreadFiber for 1000000 items
DedicatedThreadFiber.NumThreads: 8
System.Threading.ThreadPool
00:00:00.3950000
Helios.Concurrency.DedicatedThreadFiber
00:00:01.8900000
Comparing Helios.Concurrency.DedicatedThreadPool vs Helios.Concurrency.DedicatedThreadFiber for 10000000 items
DedicatedThreadFiber.NumThreads: 8
System.Threading.ThreadPool
00:00:03.6120000
Helios.Concurrency.DedicatedThreadFiber
00:00:23.0930000

Even with the `BlockingCollection` the `DedicatedThreadFiber` holds its own compared to the CLR threadpool - the biggest differences in performance really show up when the number of queued items begins to vastly exceed the size of the `BlockingCollection`'s upper limit (250,000 as of my most recent PR.)

I increased the upper limit from 25,000 to 250,000 and the increase in memory consumption was trivial (~5mb) and it resulted in shaving about 7 seconds off of the 10,000,000 item benchmark. I could probably increase it again.

### Conclusion

My recommendation for the short run is to see if there's anything that can be done with the DedicatedThreadPool to drastically improve its numbers. If there's nothing obvious there, then we should consider adding Helios as a dependency / referencing its source and using its `DedicatedThreadFiber` as the engine for the `ForkJoinDispatcher`. 

The `DedicatedThreadFiber` performance numbers are acceptable (IMHO) given that it's really meant to be used for guaranteeing available CLR threads for specific groups of actors for their exclusive use, not an improvement upon the throughout on the CLR threadpool.

**Comment by rogeralsing** [comment]
We should compare notes :-)
This is my just for fun attempt https://gist.github.com/rogeralsing/98f9c701a092fcd2925b

It does about 8 mil actions per sec.
It will not touch the manual reset event unless the queue is empty.. otherwise it will just operate on the concurrent queue.

It probably have a ton of other drawbacks, but its a super interesting topic, I want to understand this area better :-)

**Comment by rogeralsing** [comment]
cc @VesaKarvonen, so what are we doing wrong here? how can the default thread pool be so much faster than our homegrown solutions?

**Comment by VesaKarvonen** [comment]
Well, there are many factors that have surprisingly large effects on performance here.  I'll try to find time to look into this tonight.

**Comment by Aaronontheweb** [comment]
Have a release build of @rogeralsing's design on NuGet - includes error handling and other fun stuff.

https://github.com/helios-io/DedicatedThreadPool

Latest benchmarks are on the page. Package doesn't add any binaries, just the .cs file containing the implementation.
